### PR TITLE
Adds context level communication msgs between agent-cloud

### PR DESCRIPTION
**Pull request by juacker**
Adds context level communication msgs between agent-cloud

https://github.com/netdata/cloud-backend/issues/261

### CODE REVIEW

**Review by papazach — APPROVED** [review]
LGTM 👍🏼

**Review by papazach — APPROVED** [review]
👍🏼

**Review by papazach — COMMENTED** [review]
removing previous approval to discuss some points that we spotted with @juacker first.

**Review by papazach — APPROVED** [review]
final approval 😄

**Comment by underhood** [comment]
should we define `uint64 version_hash` as a type? in case we ever need to change it?

**Comment by papazach** [comment]
> should we define `uint64 version_hash` as a type? in case we ever need to change it (e.g. in one week we discover we have too many colissions and need bigger hash)?

The idea is that each context of each node instance will have its own version, and Costa brought up the idea for this version to be now() timestamp in usec.

I personally think that it would be helpful if that version/timestamp would be in msec. since we'll have to do a checksum from the Cloud side and if a node instance has more than ~11,5k contexts we might end up overflowing upon summing.

**Comment by amalkov** [comment]
Guys, why do we need to have node_id and claim_id part of the config? My understanding that it can be shared across all node instances

**Comment by papazach** [comment]
> `claim_id` and `node_id` are not part of the config, but of the message used to request configs to the agent:
> 
> ```js
> // UpdateContextConfigs command contains the list of missing context configs from the cloud to agent
> message UpdateContextConfigs {
>   // claim_id, node_id pair is used to identify the Node Instance
>   string claim_id = 1;
>   string node_id = 2;
>   // list of config hashes missing from cloud and requested from the agent
>   repeated string config_hashes = 3;
> }
> ```
> 
> But I agree, maybe they are not needed there. Given that this message is a copy&paste from the same message used to request chart configs, would there be any problem for the agent if we remove these fields @underhood, @stelfrag ?

I also agree if they are not needed let's remove them.
The respective message definition of the ChartsAndDims flow - UpdateChartConfigs - used to also include `claim_id`, `node_id` ([link](https://github.com/netdata/aclk-schemas/blob/master/proto/chart/v1/config.proto#L8)). 
If that is not needed either we could stop populating those fields.

File: proto/context/v1/context.proto [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+// ContextUpdated contains context data.
+message ContextUpdated {
+    // context id
+    string id = 1;
+    // context version is an epoch in milliseconds.
+    uint64 version = 2;
+    // first_entry, last_entry are used to set the time boudaries of the context
+    google.protobuf.Timestamp first_entry = 3;
+    google.protobuf.Timestamp last_entry = 4;
```
Comment (juacker) [inline proto/context/v1/context.proto:28]: timestamps -> uint64 seconds

File: proto/context/v1/context.proto [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    string claim_id = 2;
+    string node_id = 3;
+}
+
+// ContextUpdated contains context data.
+message ContextUpdated {
+    // context id
+    string id = 1;
+    // context version is an epoch in milliseconds.
+    uint64 version = 2;
```
Comment (juacker) [inline proto/context/v1/context.proto:25]: this will be in seconds

**Comment by underhood** [comment]
do we also need to add:
- agent give me your latest hash
- cloud here it is
messages?

So that when a user opens the cloud we:
1. check how old the latest hash we got from this agent is
2. if older than X (e.g. 1 hour)
3. request hash from agent
4. if hashes matched done else ask for the whole list

**Comment by juacker** [comment]
I've been digesting what we talked yesterday and I think that as we're introducing a delayed task for calculating the version hash on the cloud, we can simplify the `ContextsUpdated` message by removing the `version_hash` and `created_at` fields.

The idea would be that when running the delayed task, the version hash would be calculated in the cloud and sent to the agent, along with a `max_version` parameter to indicate the latest version we have among all the contexts. The agent would then compare it with its own values, and if they don't match, it could send a `ContextsSnapshot`, if they match, it will continue sending `ContextsUpdated` when needed as normal.

The way to send the `version_hash` and `max_version` to the agent could be through the `StartStreamingContexts` that we already use when the agent connects, or we could even rename this message to something like `ContextsCheckpoint` to make it more generic.

With this method, we avoid the agent having to calculate the version_hash in each `ContextsUpdated` message sent to the cloud, thus reducing the load on the agent and the bandwidth used.

I think this method could be used/handy on any needed event, like when a user starts using the cloud, etc.

In this diagram I try to summarise what the mechanism would be like:

Captura desde 2022-07-09 16-05-36

**Comment by underhood** [comment]
Although at first I thought this is great optimization, upon thinking about it my worry is in case there are regular updates we will never send the checkpoint message and thus potentially be in inconsistent state we don't know about forever.

**Comment by juacker** [comment]
Sending an updated proposal that is kind of a mix of the last two and @papazach and me think fullfill our requirements:


## proto
- `StartStreamingContexts` is renamed to `ContextsCheckpoint`
## Agent side
- The agent will send us the `version_hash` and `created_at` fields on `ContextsUpdated` message.

## Cloud side
- The cloud will create a delayed task to check version hash
- On every `ContextsUpdated` message received we will update the delayed task with the latest `version_hash` received and update its execution time.
- Once delayed task is triggered, we will calculate cloud version hash and will compare with the last one received from the agent. If they don't match we will send a `ContextsCheckpoint` message to the agent with ours.
- Delayed task can be configured with a max execution time to ensure we will execute it, no matter if an agent is updating very frecuently.
- Calculating the version hash on the cloud only by the delayed task instead of on every message receive, will reduce DB load on the cloud.

In this diagram we sumarize this flow:

image

Please add your comments/concerns. If we agree, please approve.

cc: @underhood @stelfrag @ktsaou @amalkov @ralphm

**Comment by juacker** [comment]
@underhood @papazach are you ok if I merge this or should I wait?

**Comment by papazach** [comment]
> @underhood @papazach are you ok if I merge this or should I wait?

I'm ok from my side to merge 👍🏼

**Pull request merged**